### PR TITLE
Update metadata plugins

### PR DESCRIPTION
**Pull request by tkdchen**
* CLOUDBLD-8265

For details, please refer to each commit message.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] Python type annotations added to new code
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [n/a] Changes to metadata also update the documentation for the metadata
- [n/a] Pull request has a link to an osbs-docs PR for user documentation updates
- [n/a] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/utils/koji.py [hunk 1]
```diff
@@ -322,7 +324,7 @@ def get_buildroot():
         },
         'container': {
             'type': 'none',
-            'arch': os.uname()[4],
+            'arch': arch or os.uname().machine,
```
Comment (rcerven) [inline atomic_reactor/utils/koji.py:328]: I think we should just use 'host_arch'

os.uname vs platform.processor give the same results for ppc,x86,s390x, and differ only for arm

arm64 vs aarch64

so I'd choose just one of os.uname or platform.processor as default

File: atomic_reactor/inner.py [hunk 2]
```diff
@@ -345,6 +345,14 @@ class ImageBuildWorkflowData(ISerializer):
     image_id: Optional[str] = None
     parent_images_digests: Dict[str, Dict[str, str]] = field(default_factory=dict)
 
+    # List of output files that are uploaded to Brew/Koji
+    # Each element is a two-strings list, local_filename and dest_filename. E.g.
+    # [
+    #     ["/path/to/data.json", "metadata.json"],
+    #     ["/path/to/build.log", "x86_64-build.log"],
+    # ]
+    cg_output_files: List[List[str]] = field(default_factory=list)
```
Comment (rcerven) [inline atomic_reactor/inner.py:354]: can we name it differently ?

brew_upload_files    or something more clear

File: atomic_reactor/plugins/post_koji_import.py [hunk 3]
```diff
@@ -440,127 +440,132 @@ def get_build(self):
...
+        remote_source_file_outputs, kojifile_components = get_maven_metadata(self.workflow.data)
 
         # add maven components alongside RPM components
         for worker_output in output:
             if worker_output['type'] == 'docker-image':
                 worker_output['components'] += kojifile_components
 
         # add remote sources tarballs and remote sources json files to output
-        for remote_source_output in [
+        for item in [
```
Comment (rcerven) [inline atomic_reactor/plugins/post_koji_import.py:482]: call it more meaningful

File: atomic_reactor/inner.py (hunk 2 above)
Comment (rcerven) [reply atomic_reactor/inner.py:354]: also would be nice to have list of dicts or named tuples

File: atomic_reactor/plugins/post_koji_import.py [hunk 4]
```diff
@@ -192,7 +192,7 @@ def set_media_types(self, extra):
 
         # Append media_types from verify images
         media_results = self.workflow.data.postbuild_results.get(PLUGIN_VERIFY_MEDIA_KEY)
-        if media_results:
+        if media_results and not isinstance(media_results, Exception):
```
Comment (rcerven) [inline atomic_reactor/plugins/post_koji_import.py:195]: verify media plugin will run, and it should be required now: https://github.com/containerbuildsystem/atomic-reactor/blob/master/atomic_reactor/tasks/binary.py#L59

because now media_types depend on it, so if verify media plugin fails, build should fail as well, as we shouldn't have empty media_types in brew

File: atomic_reactor/utils/koji.py (hunk 1 above)
Comment (lkolacek) [reply atomic_reactor/utils/koji.py:328]: +1 for choosing one as default, I believe there is also difference between x86_64 vs amd64 representation

File: atomic_reactor/plugins/post_koji_import.py [hunk 5]
```diff
@@ -133,16 +133,20 @@ def __init__(self, workflow, blocksize=None, poll_interval=5, userdata=None):
...
+
+        :return: a mapping from platform to metadata mapping. e.g. {"x86_64": {...}}
+        """
+        metadatas = self.workflow.data.postbuild_results.get(
+            PLUGIN_GATHER_BUILDS_METADATA_KEY, {}
         )
-        if not worker_metadatas:
+        if not metadatas:
             self.log.warning(
                 "No fetched worker metadata is found. Check if %s plugin ran already.",
```
Comment (hjmodi) [inline atomic_reactor/plugins/post_koji_import.py:147]: Should probably update this message as well.

File: atomic_reactor/plugins/post_koji_import.py [hunk 6]
```diff
@@ -157,7 +161,7 @@ def _iter_work_metadata_outputs(
         :type _filter: dict[str, any] or None
         :return: an iterator that yields a tuple in form (platform, output).
         """
-        for worker_platform, metadata in self._worker_metadatas.items():
+        for worker_platform, metadata in self._builds_metadatas.items():
```
Comment (hjmodi) [inline atomic_reactor/plugins/post_koji_import.py:165]: how about `image_platform` or `build_platform` for clarity?

File: tests/plugins/test_gather_builds_metadata.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                "koji_upload_dir": "path/to/upload",
+            }
+        }],
+    )
+    with pytest.raises(PluginFailedException, match="No enabled platforms"):
+        runner.run()
+
+
+@patch("koji.ClientSession", new=MockedClientSession)
+def test_fetch_worker_plugin(workflow):
```
Comment (hjmodi) [inline tests/plugins/test_gather_builds_metadata.py:52]: probably a good idea to change this to current plugin name as well

File: atomic_reactor/plugins/post_koji_import.py (hunk 6 above)
Comment (tkdchen) [reply atomic_reactor/plugins/post_koji_import.py:165]: :sweat_smile: You are right.

File: atomic_reactor/utils/koji.py (hunk 1 above)
Comment (rcerven) [reply atomic_reactor/utils/koji.py]: when I was checking brew medata.json I've seen diff only for aarch64

File: atomic_reactor/plugins/post_gather_builds_metadata.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
+"""
+Copyright (c) 2017, 2019 Red Hat, Inc
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_gather_builds_metadata.py:2]: Should be updated

File: atomic_reactor/plugins/post_gather_builds_metadata.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+    Each metadata mapping follows the format of Koji Content Generator Metadata.
+    This plugin ensures it has keys ``metadata_version``, ``buildroots`` and
+    ``output``.
+    """
+
+    key = PLUGIN_GATHER_BUILDS_METADATA_KEY
+    is_allowed_to_fail = False
+
+    args_from_user_params = map_to_user_params("koji_upload_dir")
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_gather_builds_metadata.py]: If I recall correctly, this user param used to be generated by the orchestrator plugin. We will not have it in user params anymore, but maybe we can just generate it here. I think another plugin uses the directory as well, we should make sure it's the same directory for both.

File: atomic_reactor/plugins/post_gather_builds_metadata.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        """
+        Build the metadata needed for importing the build
+
+        :return: tuple, the metadata and the list of Output instances
+        """
+        pullspec_image = self._determine_image_pullspec()
+        buildroot = get_buildroot(platform)
+        output_files, _ = get_output(workflow=self.workflow, buildroot_id=buildroot['id'],
+                                     pullspec=pullspec_image, platform=platform,
+                                     source_build=False)
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_gather_builds_metadata.py:64]: I think we need to use the platform-specific pullspec here

File: atomic_reactor/inner.py [hunk 11]
```diff
@@ -345,6 +345,14 @@ class ImageBuildWorkflowData(ISerializer):
     image_id: Optional[str] = None
     parent_images_digests: Dict[str, Dict[str, str]] = field(default_factory=dict)
 
+    # List of output files that are uploaded to Brew/Koji
+    # Each element is a two-strings list, local_filename and dest_filename. E.g.
+    # [
+    #     {"local_filename": "/path/to/data.json", "dest_filename": "metadata.json"},
+    #     {"local_filename": "/path/to/build.log", "dest_filename": "x86_64-build.log"},
+    # ]
+    koji_upload_files: List[Dict[str, str]] = field(default_factory=list)
```
Comment (chmeliik) [inline atomic_reactor/inner.py:354]: This could also be a good way to save logs from `bsi` (source) / `podman` (binary) as a replacement for the removed BuildResult.logs :+1:

File: atomic_reactor/plugins/post_koji_import.py [hunk 12]
```diff
@@ -442,38 +442,54 @@ def get_build(self):
...
+
+        # Collect log files
         osbs_logs = OSBSLogs(self.log, get_platforms(self.workflow.data))
-        output_files = [add_log_type(add_buildroot_id(md, buildroot_id))
-                        for md in osbs_logs.get_log_files(self.osbs, self.pipeline_run_name)]
+        output_log_files = [
+            add_log_type(add_buildroot_id(md, buildroot_id))
+            for md in osbs_logs.get_log_files(self.osbs, self.pipeline_run_name)
+        ]
+        for remote_source_file_output in output_log_files:
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_koji_import.py:475]: Why is it called `remote_source_file_output`? They're just log files, aren't they? :thinking:

File: atomic_reactor/plugins/post_koji_import.py [hunk 13]
```diff
@@ -573,14 +590,17 @@ def run(self):
...
-        finally:
-            for output in output_files:
-                if output.file:
-                    output.file.close()
+        server_dir = self.get_server_dir()
+        koji_metadata = self.combine_metadata_fragments()
+
+        if is_scratch_build(self.workflow):
+            self.upload_scratch_metadata(koji_metadata, server_dir)
+            return
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_koji_import.py:588]: Previously, the plugin did this before checking the task state. I don't know if that is important, but maybe we want to keep it that way?

**Review by chmeliik — COMMENTED** [review]
Nice! I like the decision to keep some of the metadata gathering in gather_builds_metadata, that makes the changes look simpler than I thought they were going to be :+1:

File: atomic_reactor/plugins/post_gather_builds_metadata.py (hunk 9 above)
Comment (chmeliik) [reply atomic_reactor/plugins/post_gather_builds_metadata.py]: Addressed in later commit, but this is still relevant for `KojiImportPlugin.get_server_dir()` - now it should generate the upload dir instead of getting it from the orchestrator plugin

File: atomic_reactor/utils/koji.py [hunk 14]
```diff
@@ -400,25 +407,24 @@ def get_maven_metadata(workflow_data):
...
-def get_image_components(workflow):
+def get_image_components(workflow, platform: str) -> List[Dict[str, Union[str, int]]]:
     """
     Re-package the output of the rpmqa plugin into the format required
     for the metadata.
     """
-    output = workflow.data.image_components
-    if output is None:
+    components = workflow.data.postbuild_results.get(PostBuildRPMqaPlugin.key)
+    if components is None:
```
Comment (hjmodi) [inline atomic_reactor/utils/koji.py:421]: I don't think the result of `rpmqa` will ever be `None`, at most it'll be something likes `{platfrom: None}` since we are using `build_dir.for_each_platfrom` to return the results

File: tests/plugins/test_gather_builds_metadata.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
+"""
+Copyright (c) 2017 Red Hat, Inc
```
Comment (hjmodi) [inline tests/plugins/test_gather_builds_metadata.py:2]:
```suggestion
Copyright (c) 2017-2022 Red Hat, Inc
```

**Review by hjmodi — APPROVED** [review]

**Comment by hjmodi** [comment]
LGTM, the code is much more readable now! :tada: 

Just a couple minor comments

File: atomic_reactor/utils/koji.py (hunk 14 above)
Comment (tkdchen) [reply atomic_reactor/utils/koji.py:421]: Hi @hjmodi, thanks for your comment. As you can see the following `logger.error`, this check is checking whether `PostBuildRPMqaPlugin` plugin has run. If not, no result in the `postbuild_results`.

File: atomic_reactor/plugins/post_koji_import.py [hunk 16]
```diff
@@ -15,20 +15,21 @@
...
+import koji_cli.lib
+
 from atomic_reactor.config import get_koji_session, get_openshift_session
 from atomic_reactor import start_time as atomic_reactor_start_time
 from atomic_reactor.plugin import PostBuildPlugin
 from atomic_reactor.plugins.pre_add_help import AddHelpPlugin
 from atomic_reactor.source import GitSource
-from atomic_reactor.plugins.build_orchestrate_build import (get_worker_build_info,
-                                                            get_koji_upload_dir)
+from atomic_reactor.plugins.build_orchestrate_build import get_worker_build_info
```
Comment (chmeliik) [inline atomic_reactor/plugins/post_koji_import.py:25]: We should not be importing this anymore. Should be fine to address this in a separate PR.

File: tests/utils/test_koji.py [hunk 17]
```diff
@@ -416,3 +428,146 @@ def test_with_notable(self, totalsize, step, notable, expected_times):
...
+    if has_export_operator_manifests:
+        assert len(output) == 2
+    else:
+        assert len(output) == 1
+
+    image_metadata = output[0].metadata
+
+    # Assert these image metadata firstly, then remove them and assert the
+    # rest. So, no need to mock anything for get_image_output.
+    assert 'docker-image-None.x86_64.tar.gz' == image_metadata.pop('filename')
```
[the remark this answers is not in the document]
Comment (tkdchen) [reply tests/utils/test_koji.py:559]: It is the `image_id` (`workflow.data.image_id`), which will be fixed in next step.

File: tests/utils/test_koji.py [hunk 18]
```diff
@@ -416,3 +428,146 @@ def test_with_notable(self, totalsize, step, notable, expected_times):
...
+    workflow.conf.conf = {
+        'registries': [
+            {'url': 'https://registry.host/', 'insecure': False},
+        ],
+    }
+    # Mock get manifest digests
+    # What would happen if there is no both v2 and oci digest?
+    image_manifest_digest = ManifestDigest(
+        {'oci': 'oci-1234'} if no_v2_digest else {'v2': '1234'}
+    )
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/utils/test_koji.py:482]: ah, right :smile:

File: tests/utils/test_koji.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/utils/test_koji.py:559]: Ack, seems reasonable

File: tests/utils/test_koji.py [hunk 19]
```diff
@@ -416,3 +428,146 @@ def test_with_notable(self, totalsize, step, notable, expected_times):
...
+        'arch': platform,
+        'type': 'docker-image',
+        'components': [],
+        'extra': {
+            'image': {'arch': platform},
+            'docker': {
+                'id': None,
+                'repositories': expected_repositories,
+                'layer_sizes': layer_sizes,
+                'tags': sorted([primary_image.tag, unique_image.tag]),
```
[the remark this answers is not in the document]
Comment (tkdchen) [reply tests/utils/test_koji.py:531]: I'll address in next step as well. Will open another PR.

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

Were you going to squash the commits or keep all 18? I don't really mind either way

File: atomic_reactor/utils/koji.py (hunk 14 above)
Comment (hjmodi) [reply atomic_reactor/utils/koji.py:423]: hmm, I do understand that. It'll only not run for source build in which case this function will not be called. So I don't think we need to check for that. It might lead to confusion. It'll always run for the binary build, if no RPMs are found it'll return a `{plafrom: None}` dictionary

**Comment by tkdchen** [comment]
The build-osbs-components is still unavailable. I'm going to merge this pr for now.

@chmeliik Keep these 18 commits :sweat_smile:

File: atomic_reactor/utils/koji.py (hunk 14 above)
Comment (tkdchen) [reply atomic_reactor/utils/koji.py:423]: My understand, the original decision of this check is based on the plugins relationship rather than what kind of data is set to the `image_components`. For now, I'd like keep this original behavior unchanged. Maybe, `log.warning` is more proper than `log.error`.
